collector/collector.go

```diff
@@ -0,0 +1,499 @@
+	}
+
+	filteredYAML, err := filter(dotToYAML)
+	if err != nil {
+		return &result, err
+	}
+
+	err = yaml.Unmarshal(filteredYAML, &result)
+	if err != nil {
+		return &result, err
+	}
+
+	return &result, nil
+}
+
+// ConfigURL returns the value of config_url if set or empty string otherwise.
+func (c Config) ConfigURL() string {
+	if val, hasKey := c["config_url"]; hasKey {
+		if s, isString := val.(string); isString {
+			return s
+		}
+	}
+
+	return ""
+}
+
+func fetchRemoteConfig(url string) (*Config, error) {
+	var body []byte
+	result := &Config{}
+
+	err := retry.Do(
+		func() error {
+			resp, err := http.Get(url)
+			if err != nil {
+				return err
+			}
+			if resp.StatusCode != http.StatusOK {
+				return fmt.Errorf("unexpected status: %d", resp.StatusCode)
+			}
+			defer resp.Body.Close()
+
+			body, err = io.ReadAll(resp.Body)
+			if err != nil {
+				return err
+			}
+
+			return nil
+		}, retry.Delay(time.Second), retry.Attempts(3),
+	)
+
+	if err != nil {
+		// TODO: improve logging
+		fmt.Printf("WARNING: Couldn't fetch config_url: %s", err)
+		return result, nil
+	}
+
+	if !HasValidHeader(string(body)) {
+		// TODO: Print a warning when we implement proper logging
+		fmt.Println("No valid header in remote config: %w", err)
+		return result, nil
+	}
+
+	if err := yaml.Unmarshal(body, result); err != nil {
+		return result, fmt.Errorf("could not unmarshal remote config to an object: %w", err)
+	}
+
+	return result, nil
+}
+
+func HasValidHeader(data string) bool {
+	header := strings.SplitN(data, "\n", 2)[0]
+
+	// Trim trailing whitespaces
+	header = strings.TrimRightFunc(header, unicode.IsSpace)
+
+	// NOTE: we also allow "legacy" headers. Should only allow #cloud-config at
+	// some point.
+	return (header == DefaultHeader) || (header == "#kairos-config") || (header == "#node-config")
+}
+
+func (c Config) Query(s string) (res string, err error) {
+	s = fmt.Sprintf(".%s", s)
+
+	var dat map[string]interface{}
+	var dat1 map[string]interface{}
+
+	yamlStr, err := c.String()
+	if err != nil {
+		panic(err)
+	}
+	// Marshall it so it removes the first line which cannot be parsed
+	err = yaml.Unmarshal([]byte(yamlStr), &dat1)
+	if err != nil {
+		panic(err)
+	}
+	// Transform it to json so its parsed correctly by gojq
+	b, err := json.Marshal(dat1)
+	if err != nil {
+		panic(err)
+	}
+	if err := json.Unmarshal(b, &dat); err != nil {
+		panic(err)
+	}
+
+	query, err := gojq.Parse(s)
+	if err != nil {
+		return res, err
+	}
+
+	iter := query.Run(dat) // or query.RunWithContext
+	for {
+		v, ok := iter.Next()
+		if !ok {
+			break
+		}
+		if err, ok := v.(error); ok {
+			return res, fmt.Errorf("failed parsing, error: %w", err)
+		}
+
+		dat, err := yaml.Marshal(v)
+		if err != nil {
+			break
+		}
+		res += string(dat)
+	}
+	return
+}
+
+// FilterKeys is used to pass to any other pkg which might want to see which part of the config matches the Kairos config.
+func FilterKeys(d []byte) ([]byte, error) {
+	cmdLineFilter := Config{}
```

Comment:
The problem is that this `Config` is making reference to the collector.Config, which is a `map[string]interface{}` which means it will never filter anything. Whiel the one in the kairos-agent is a `config.Config` (I know, super confusing :facepalm: ) and that one has a proper struct so it will only catch matching values